keymanapp/keyman
# feat(oem/fv/android): Add keyboard and dictionary controls

This adds a feature to the FirstVoices Android app for keyboard and dictionary management. 

Several of these "new" files are direct copies from the Keyman for Android app (kmapro) to handle the kmp download, installation, and settings menus. The dictionary kmp installations are hard-coded to "silent" install (doesn't display help documentation).

Changed workflow:
The current FV app has checkboxes in the keyboard list to add/remove keyboards during setup. These have been replaced with forward arrows that go to keyboard settings menu. Enabling the toggle there will add the keyboard and check for associated dictionary to download. If there's a matching dictionary from api.keyman.com, the kmp will download in the background and silently install.

This also incorporates changes from #2255 and #4696 to disable suggestions when entering password field, and re-enabling when exiting.

Tagging @madebybridget

## Screenshots

Region menu - no change
![regions](https://user-images.githubusercontent.com/7358010/155473940-0d7a654b-f269-407a-8cd1-95e8a02ae21f.png)

BC Coast menu - left blue checkbox shows currently installed keyboards; checkboxes replaced with forward arrows
![bc coast](https://user-images.githubusercontent.com/7358010/155473843-6103509b-5527-4978-8019-47ceb3d280f6.png)

Haisla keyboard settings
![haisla](https://user-images.githubusercontent.com/7358010/155474039-48299070-50f3-489b-a0e4-e2e7eb2d406a.png)

Sencoten keyboard settings (dictionary name updates after it's installed)
![sencoten](https://user-images.githubusercontent.com/7358010/155474059-1137b56d-4088-4900-a86e-aebec60c3087.png)

Model Picker menu (copied from kmapro)
![str dictionaries](https://user-images.githubusercontent.com/7358010/155474166-20074067-c354-4d2d-80d5-d549c8df05ac.png)

str dictionary info (also copied)
![dictionary info](https://user-images.githubusercontent.com/7358010/155474205-4419ba29-e2f8-46e7-93d8-3997457c0252.png)

## User Testing
* **TEST_FV_ANDROID**
1. On an Android device/emulator, install and load the PR build of FV Android
2. When the FirstVoices app loads, click the "Select keyboards" button
3. Select "BC Coast" region --> forward arrow --> "Sencoten"
4. When the Sencoten keyboard setting menu is displayed, toggle the enable button to add the keyboard
5. Verify a notification appears on the bottom about checking for an associated dictionary to install
6. Allow time for the associated dictionary to download and install in the background. This may take minutes depending on network connection
7. Click on the Dictionary row
8. Verify the "Straits Salish (Latin) dictionaries" menu appears with SENCOTEN dictionary and a blue checkbox appear
9. Click the forward arrow
10. Verify the Sencoten dictionary info menu appears and shows
    * version
    * help link
    * Uninstall dictionary
11. Click the back arrow several times to return to the main menu
12. Click the "Setup" button to finish setting up FirstVoices as a system keyboard
13. Launch a separate app (e.g. Chrome)
14. In the Chrome browser searchbox, select the FirstVoices keyboard
15. Verify the Sencoten keyboard and dictionary appear
16. Return to the FV app and navigate to the Sencoten Settings menu
17. Disable predictions and verify corrections are also disabled
18. Return to Chrome and verify the Sencoten suggestions no longer appear
19. Return to the FV app and enable predictions
20. Return to Chrome and verify the Sencoten suggestions appear
21. Return to the FV app and navigate to the Sencoten dictionary info menu
22. Click "Uninstall dictionary"
23. Return to Chrome and verify the Sencoten suggestions no longer appear

* **TEST_PASSWORD_FIELD** - Tests that suggestions are disabled on password field and re-enabled leaving it
1. On an Android device/emulator, install and load the PR build of FV Android
2. Following the keyboard setup to install "BC Coast" region --> "Sencoten" along with its associated dictionary
3. Finish the FV setup to enable FV as a system keyboard
4. Launch Chrome and go to https://darcywong00.github.io/examples/form
5. Select the "Visible text field" and verify Sencoten suggestions appear as you type
7. Select the "Search field" and verify Sencoten suggestions appear as you type
8. Select the "Password field" and verify Sencoten suggestions are disabled as you type
9. Go back to "Visible text field" and verify Sencoten suggestions appear as you type
